Cleanup the stats router code

jsnajdr opened the pull request:
The Stats router code has accumulated a lot of cruft over the years, and this PR tries to clean it up. The material benefit is that the `stats` section loses 40% of its JavaScript weight, mainly due to removed dependency on the `activity` section (after moving the redirect handler).

The changes are:
- remove the unit test for the router, as it's an example of not-useful test that only duplicates code: "the code calls `page( a, b, c )`, so let's mock `page` and verify it's been called with `a`, `b`, `c`". If I understand it correctly, @scinos created the test to test LUX tracking, but that has been removed since. During this refactoring, my experience has been that the test creates more work for me and doesn't help me verify correctness.
- remove the `manage/stats` feature flag that's been always `true` for several years
- migrate the `stats/controller` to individual ESM exports instead of `export default { a, b, c }`, which was probably a relict of CJS to ESM migration
- instead of defining routes for `/stats/day`, `/stats/week`, etc. individually, catch all valid periods in a `:period(day|week)` matcher.
- move the `redirectToActivity` handler from the `activity` module to `stats`. Independent sections should import each other just to figure out the other section's URL. The removed dependency makes the `stats` section chunk 40% smaller.
- rename `trackingPage` (no more tracking there) to `statsPage`, a convenient helper

**How to test:**
Verify that stats work exactly as they did before, and that this PR is a pure refactoring that doesn't change any behavior at all.

jsnajdr committed:
Remove the manage/stats feature flag that is always on

jsnajdr committed:
Stats Controller: use individual ESM exports for functions

jsnajdr committed:
Stats Router: use validPeriod matcher to define routes

jsnajdr committed:
Move the redirectToActivity handler to the stats module

jsnajdr committed:
Simplify the trivial /stats to /stats/day redirect

jsnajdr committed:
Rename trackedPage (tracking is gone) to statsPage

scinos commented:
I strongly think we should keep the test `client/my-sites/stats/test/index.js`. If this PR is only a refactorying, the test should pass without changes, because it tests that we define a set of routes and the middlewares+handlers for those routes.

I don't agree with the statement

> as it's an example of not-useful test that only duplicates code: "the code calls page( a, b, c ), so let's mock page and verify it's been called with a, b, c".

The test is not not-useful as it can (and should) be used to prove that after this refactoring, the exposed routes have not changed. It it is true that essentially it just mocks `page` and verifies it has been called with a set of arguments. But given that `client/my-sites/stats/index.js` calls that API many many times with different sets of arguments, the test will guarantee that we don't change any of those calls by accident when we do a refactor.

How did the test created more work for you? Can we change the test to reduce that?

scinos commented:
How good is the test coverage for `client/my-sites/stats/controller.jsx `? I don't see any unit test, not sure about e2e.

The change feels too big for me to comfortably approving it. I'd suggest adding unit test to assert current behaviour before refactoring it.

scinos reviewed (dismissed):
Restore deleted unit test

jsnajdr committed:
Fix the stats router unit test

jsnajdr committed:
Stats mock controller: use ESM imports, make mocks unique

Before this patch, all `statsController.*` handlers were `undefined`, so
swapping, e.g., `statsController.insights` for `statsController.post` still
passed the tests.

jsnajdr commented:
> How did the test created more work for you?

I removed the commit that deletes the test and fixed it instead in https://github.com/Automattic/wp-calypso/pull/45062/commits/a201fb2d4a3eba642ed2b69178068eab0964ff9e

To make it work, I had to duplicate the code changes that I did to the actual controller:
- merging the `/stats/day`, `/stats/week` routes into single `/stats/:period(day|week)`
- importing the `redirectToActivity` handler from another module
- removing the mocks for modules that the router no longer uses (`lib/route`, `config`)

Note that the test doesn't check any _behavior_ at all. Only that the `page` function is called with the arguments that are written in the code. In other words, the test is not testing if the router is routing correctly, but rather that the JS interpreter passes function arguments correctly. That's not a very valuable thing to test.

And the test does even this argument-checking poorly. See the improvement I did in https://github.com/Automattic/wp-calypso/pull/45062/commits/3b7ba79f31497ac6f9383c133cbde6f3dd08870c. Before that patch, the `statsController` module was mocked in a way where all exports were `undefined`. Therefore, if I swap `statsController.insights` for `statsController.post` in the code or in the test, the test continues passing. I changed the mock to make each of the exports an unique instance of `jest.fn()`.

> the test will guarantee that we don't change any of those calls by accident when we do a refactor.

The test achieves that by duplicating the code. If I change `page( 1 )` to `page( 2 )` in the code, I have to change `1` to `2` in the unit tests, too. That's all. Good unit tests are supposed to test behavior, not the internals of the implementation.

> Can we change the test to reduce that extra work?

Yes, we can delete it and write something else from scratch 🙂 Something that tests the actual routing. I.e., that when the stats router is asked to render `/stats/day`, for example, the handler chain that is called produces a page for daily stats.

This would be arguably a rather complex test, as the router code is the kind of code that _wires up_ several modules together. It's not very amenable to unit testing. We have a set of e2e tests that test the stats section, and I think that's the right way to test the router. I don't see much value in attempts to unit test the router.

scinos commented:
Looks like we fundamentally disagree of what that test is doing.

From my point of view, that test is not testing that the router is routing correctly. It is testing that an external module (aka the router) is configured correctly. From that point of view, `client/my-sites/stats/index.js` is just a glorified configuration file for `page.js`. The test is testing it is configured correctly, not that it actually works because that would mean testing the behaviour of `page.js` as well, therefore it won't be an unit test anymore. I agree we can introduce other types of test to verify the actual routing behavior _as well_. That doesn't invalidate this unit test.

> > the test will guarantee that we don't change any of those calls by accident when we do a refactor.

> The test achieves that by duplicating the code. If I change page( 1 ) to page( 2 ) in the code, I have to change 1 to 2 in the unit tests, too. 

That's 100 correct and a good thing. If you change `page(1)` to `page(2)` by _accident_, the test will prevent that. If you change it _on purpose_, you have to change the test because you have changed how we configure `page.js` router. In other words, if you change how you use an external module, you have to change the unit test. That's expected for an unit test.

> To make it work, I had to duplicate the code changes that I did to the actual controller:

But why? Wasn't the original test passing? What was failing?

>  Good unit tests are supposed to test behavior, not the internals of the implementation.

The exact goal of this test is to ensure the set of routes + handlers doesn't change by accident even if we refactor `client/my-sites/stats/index.js` to remove duplication (i.e. change the implementation of that module). This test is testing the behaviour of the module, form the point of view that all the module is doing is configuring the router in a specific way.

jsnajdr commented:
> That's 100 correct and a good thing. If you change page(1) to page(2) by accident, the test will prevent that.

I think this kind of test is a horrible one that only artificially improves test coverage metrics and is not really testing anything. Let's agree to disagree here 🙂

We have plenty of such unit tests, for sure. A typical example are Redux action creators:

Code:
```js
function setBlog( blogId ) {
  return { type: SET_BLOG, blogId };
}
```

Test:
```js
expect( setBlog( 123 ) ).toEqual( { type: SET_BLOG, blogId: 123 } );
```

Maybe that prevents some accidents, but we could achieve the same result by keeping two copies of each source file and testing if they're identical:

Source folder:
```js
stats/controller.js
stats/controller.test.js
```

Universal unit test that achieves 100% coverage instantly:
```js
sources.forEach( source => {
  expect( readFile( source ) ).toEqual( readFile( addTestSuffix( source ) ) );
} );
```
